polynote-server/src/main/scala/polynote/server/repository/fs/LocalFilesystem.scala

```diff
@@ -22,9 +23,55 @@ class LocalFilesystem(maxDepth: Int = 4) extends NotebookFilesystem {
     content <- readBytes(is).ensuring(ZIO.effectTotal(is.close()))
   } yield new String(content.toArray, StandardCharsets.UTF_8)
 
+  /**
+    * Writes the content to the given path "atomically" – at least, in a way where the existing file at that path (if
+    * present) won't be unrecoverable if writing fails for some reason.
+    */
+  private def writeAndSwap(path: Path, content: Array[Byte]) =
+    try {
+      // first, try creating the file in case it doesn't exist. If it doesn't, go ahead and write directly to it.
+      // it's not great to use an exception for normal flow control, but I don't think there's another way to avoid a
+      // race condition here (e.g. if we checked that the file doesn't exist, but then it does by the time we go to
+      // write to it)
+      Files.createFile(path)
+      Files.write(path, content)
```

Comment:
Could we remove the race condition using `Files.write(path, content, CREATE_NEW)`?

Reply:
Hey, good idea. I think so! But still has the same issue wrt to the comment – we can't avoid abusing the exception for normal, expected flow control.

Reply:
Yeah that's too bad 😞